### ISSUE TITLE
`quotes` never flags nested quotes

### ISSUE THREAD

**Issue by IanVS** [eslint/eslint#3646]
Given the following, shouldn't there be an error, since the `"avoid-escape"` option is not set?

```js
/*eslint quotes: [2, "double"]*/

var escaped = "a string containing 'single' quotes";
```

But, when I run this, I get no errors.  I checked with the Espree demo, and there is only a single literal node: `"a string containing 'single' quotes"`, and the way the rule works, the interior quotes are never examined.  I was also surprised there are no invalid test cases for nested quotes.  I can take this on if accepted.

**Comment by mysticatea** [comment]
I thought it's valid.

```js
/*eslint quotes: [2, "double"]*/
var escaped = "a string containing 'single' quotes"; // valid.
var escaped = 'a string containing "double" quotes'; // invalid. it should enclose with double quotes and it should escape double quotes in the string.
```

```js
/*eslint quotes: [2, "double", "avoid-escape"]*/
var escaped = "a string containing 'single' quotes"; // valid.
var escaped = 'a string containing "double" quotes'; // valid single quotes to avoid escaping double quotes in the string.
```

**Comment by IanVS** [comment]
I'm not sure I understand the rule the same way.  Without `"avoid-escapes"`, I think ALL quotes should follow the rules specified, and nested quotes must be escaped rather than using the alternate style.

Maybe that's just my assumption, that the outer quotes should always follow the rules, and the inner quotes should either be escaped, or the `"avoid-escapes"` option should be used to enable alternate style within.

**Comment by mysticatea** [comment]
The option seems to be added in this issue: #199

**Comment by ilyavolodin** [comment]
@IanVS It's slightly different. Without `avoid-escape` this rule will accept escaped quotes as well as another type of quotes inside strings. With it, it will flag any escaped quotes.

**Comment by IanVS** [comment]
> With it, it will flag any escaped quotes.

Are you sure?  The following gives me no problems:

```js
/*eslint quotes: [2, "double", "avoid-escape"]*/

var escaped = "a string containing \"escaped\" quotes";
```

**Comment by IanVS** [comment]
This also passes:

```js
/*eslint quotes: [2, "double", "avoid-escape"]*/

var escaped = 'a string containing \'single" quotes';
```

So, it seems that the only condition when using `"avoid-escape"` is that one of the correct kinds of quotes is found somewhere in the string.  It allows the non-standard quotes to be used on the outermost positions, that's it.

There's currently no way to force the use of escaping to maintain a single style of quotes.  Not sure if that's a bug or not, but it seems like something which should be possible to enforce.

**Comment by ilyavolodin** [comment]
Agree. Latest examples seems like a bug

**Comment by nzakas** [comment]
Just to close the loop: this rule is about quotes used to define strings and nothing else, so no "inner quote" checking is done unless the "avoid-escape" option is used. Even then, it only makes sure that you don't escape a quote character because you used it in the string.

**Comment by IanVS** [comment]
> Even then, it only makes sure that you don't escape a quote character because you used it in the string.

This may be the intent, but as I showed [above](https://github.com/eslint/eslint/issues/3646#issuecomment-137647092), it is not the current behavior.

The current behavior of `"avoid-escape"` is to allow the opposite quote style around a string, as long as the specified quote style is found within the string.

**Comment by nzakas** [comment]
Yes, as @ilyavolodin mentioned, that's a bug. We're in agreement on that.

**Comment by nzakas** [comment]
Here's the problem:

```js
if (!isValid && avoidEscape) {
    isValid = isSurroundedBy(rawVal, settings.alternateQuote) && rawVal.indexOf(settings.quote) >= 0;
}
```

The `>=0` is applied to the raw representation, which includes the surrounding quotes.

**Comment by nzakas** [comment]
Actually, looks like a bunch of problems. I'll work on this.

**Comment by michaelficarra** [comment]
I would prefer avoid-escape to mean that you *may* avoid escaping internal quotes by changing the delimiter, not that you *must* do so. If this is changed, my use case will be lost.

**Comment by nzakas** [comment]
Agreed. Looking at the original issue (#199), I believe the correct functionality is to add an exception so that strings pass rather than adding one so that they fail. Example:

```js
// functionality for quotes: [1, "double"]
var foo = "this is \"my\" string"  // pass
var foo = 'this is "my" string'  // warn

// functionality for quotes: [1, "double", "avoid-escape"]
var foo = "this is \"my\" string"  // pass
var foo = 'this is "my" string'  // pass, because double quotes found inside
```

**Comment by IanVS** [comment]
https://github.com/eslint/eslint/issues/3646#issuecomment-137858218 seems to be the current behavior.

My original question was about:

```js
// functionality for quotes: [1, "double"]

var foo = "this is 'my' string"  // warn or pass?
```

I'm thinking this should pass based on the above discussion.  If so, this issue can be closed.

**Comment by IanVS** [comment]
Sorry, one more question, should this result in an error?

```js
/*eslint quotes: [2, "single", "avoid-escape"]*/

var escaped = "a string containing escaped \'single\' quotes";
```

**Comment by ilyavolodin** [comment]
@IanVS first example should pass without warning. Second should fail.

**Comment by nzakas** [comment]
It seems like what we really need here is to update the docs to better explain the current behavior.

**Comment by IanVS** [comment]
@ilyavolodin 
> first example should pass without warning. Second should fail.

OK, the first example does pass, but so does the second. :-/

**Comment by ilyavolodin** [comment]
That looks like a bug, not a documentation issue. @nzakas any thoughts? Rule requires single quotes around strings, and `avoid-escape` should change that behavior, since quotes inside are already escaped.

**Comment by nzakas** [comment]
I'm having trouble tracking this conversation. I'd like to ask that when reporting additional possible bugs that we follow the convention of:

1. Here's what I did
2. Here's what happened
3. Here's what I expected to happen

The "what should this do?" line of questioning is hard to triangulate for me.

I don't believe this should be a warning:

```js
/*eslint quotes: [2, "single", "avoid-escape"]*/
var escaped = "a string containing escaped \'single\' quotes";
```

Because it is functionally the same as this:

```js
/*eslint quotes: [2, "single", "avoid-escape"]*/
var escaped = "a string containing escaped 'single' quotes";
```

That's to say, this rule only flags problems with outer quotes and `avoid-escape` just provides an exception. This rule does not enforce unescaped quotes in strings.

**Comment by michaelficarra** [comment]
:+1:

**Comment by IanVS** [comment]
> 3. Here's what I expected to happen

That's the difficult part for me in this case, as I couldn't tell what I should expect based on the docs and what was being said in this issue.  I think now we've come full circle to it being a documentation issue, as far as I can tell.

**Comment by nzakas** [comment]
Understood - even if you don't exactly know what you expect, it's still useful to report what it's actually doing as a conversation starter. 

I'll tackle the documentation.
